cpu/esp_common/esp-now/esp_now_netdev.c

```diff
@@ -200,40 +202,56 @@ static IRAM_ATTR void esp_now_recv_cb(const uint8_t *mac, const uint8_t *data, i
     }
 #endif
 
-    mutex_lock(&_esp_now_dev.rx_lock);
-    critical_enter();
+    /*
+     * The function `esp_now_recv_cb` is executed in the context of the `wifi`
+     * thread. The ISRs handling the hardware interrupts from the WiFi
+     * interface pass events to a message queue of the `wifi` thread which is
+     * sequentially processed by the `wifi` thread to asynchronously execute
+     * callback functions such as `esp_now_recv_cb`.
+     *
+     * It should be therefore not possible to reenter function
+     * `esp_now_recv_cb`. To avoid inconsistencies this is checked by an
+     * additional boolean variable . This can not be realized by a mutex
+     * because `esp_now_recv_cb` would be reentered from same thread context.
+     * If the NDEBUG macro is undefined, an assertion is used instead for
+     * debugging purposes.
+     */
+    assert(!_in_recv_cb);
+
+    if (_in_recv_cb) {
+        return;
+    }
+    _in_recv_cb = true;
 
     /*
-     * The ring buffer uses a single byte for the pkt length, followed by the mac address,
-     * followed by the actual packet data. The MTU for ESP-NOW is 250 bytes, so len will never
-     * exceed the limits of a byte as the mac address length is not included.
+     * Since it is not possible to reenter the function `esp_now_recv_cb`, and
+     * the functions netif::_ recv and esp_now_netdev::_ recv are called
+     * directly in the same thread context, neither a mutual exclusion has to
+     * be realized nor have the interrupts to be deactivated.
+     * Therefore we can read directly from the `buf` and don't need a receive
+     * buffer.
      */
-    if ((int)ringbuffer_get_free(&_esp_now_dev.rx_buf) < 1 + ESP_NOW_ADDR_LEN + len) {
-        critical_exit();
-        mutex_unlock(&_esp_now_dev.rx_lock);
+    if (_esp_now_dev.rx_len) {
         DEBUG("%s: buffer full, dropping incoming packet of %d bytes\n", __func__, len);
+        _in_recv_cb = false;
         return;
     }
 
-#if 0 /* don't printf anything in ISR */
-    printf("%s\n", __func__);
-    printf("%s: received %d byte from %02x:%02x:%02x:%02x:%02x:%02x\n",
-           __func__, len,
-           mac[0], mac[1], mac[2], mac[3], mac[4], mac[5]);
-    od_hex_dump(data, len, OD_WIDTH_DEFAULT);
-#endif
-
-    ringbuffer_add_one(&_esp_now_dev.rx_buf, len);
-    ringbuffer_add(&_esp_now_dev.rx_buf, (char*)mac, ESP_NOW_ADDR_LEN);
-    ringbuffer_add(&_esp_now_dev.rx_buf, (char*)data, len);
+    _esp_now_dev.rx_mac = (uint8_t*)mac;
+    _esp_now_dev.rx_data = (uint8_t*)data;
+    _esp_now_dev.rx_len = len;
 
+    /*
+     * Since we are not in the interrupt context, we do not have to pass
+     * `NETDEV_EVENT_ISR` first. We can call the receive function directly.
+     * But we have to unlock the mutex and enable interrupts before.
```

Comment:
which mutex?

Reply:
The comments is from  the time when a mutex was used. I forgot to change it it when I removed :wink:

Reply:
please remove then, afterwards this seems good to go

Reply:
Of course. Thank you for the review.